specs/capri/API_Framework_Refactoring.md

```diff
@@ -0,0 +1,384 @@
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// CreateVolumeSnapshotOpts is a structure which indicates all required properties
+// for creating a volume snapshot.
+message CreateVolumeSnapshotOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// DeleteVolumeSnapshotOpts is a structure which indicates all required properties
+// for deleting a volume snapshot.
+message DeleteVolumeSnapshotOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// CreateReplicationOpts is a structure which indicates all required properties
+// for creating a replication.
+// NOTE: Need to figure out how to handle more than 2 sites.
+message CreateReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+// for deleting a replication.
+// NOTE: Need to figure out how to handle more than 2 sites.
+message DeleteReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+message EnableReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+message DisableReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+// Delete ReplicationOpts is a structure which indicates all required properties
+message FailoverReplicationOpts {
+    // The message of request, required.
+    string message = 1;
+    string failoverMessage = 2;
+    // The Context
+    string context = 3;
+}
+
+// CreateVolumeGroupOpts is a structure which indicates all required
+// properties for creating a volume group.
+message CreateVolumeGroupOpts {
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+message UpdateVolumeGroupOpts{
+    // The message of request, required.
+    string message = 1;
+    repeated string addVolMessage = 2;
+    repeated string rmVolMessage = 3;
+    // The Context
+    string context = 4;
+}
+
+message DeleteVolumeGroupOpts{
+    // The message of request, required.
+    string message = 1;
+    // The Context
+    string context = 2;
+}
+
+message GenericResponse {}
+```
+
+### REST API impact
+
+None
+
+### Security impact
+
+None
+
+### Other end user impact
+
+None
+
+### Performance impact
+
+##### Pros
+Decoupling `api-server` from `controller` would make `api-server` much easier to scale out, and therefore improving
+performance when handling heavy workloads.
```

Comment:
It does allow api-server to scale out, but I don't think this improves performance as it adds another layer of indirection.

Reply:
What I mean here is that for heavy workloads, only one api-server will cause the traffic delay for the request, but as for multiple api-server modules all requests will be split and therefore the overall time would be reduced. Does that make sense?